finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala

```diff
@@ -0,0 +1,280 @@
+            HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(),
+            Http2CodecUtil.HTTP_UPGRADE_STREAM_ID
+          )
+
+          val adapter = new InboundHttpToHttp2Adapter(
+            http2Handler.connection,
+            http2Handler.decoder.frameListener
+          )
+          adapter.channelRead(ctx, upgrade.upgradeRequest.retain())
+        } finally { upgrade.release(); () }
+
+      case ev => super.userEventTriggered(ctx, ev)
+    }
+
+  // Override this to signal it will never throw an exception.
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    ctx.fireExceptionCaught(cause); ()
+  }
+
+  /**
+   * Processes all {@link Http2Frame}s. {@link Http2StreamFrame}s may
+   * only originate in child streams.
+   */
+  override def write(ctx: ChannelHandlerContext, msg: Any, promise: ChannelPromise): Unit =
+    msg match {
+      case goaway: Http2GoAwayFrame =>
+        try {
+          if (goaway.lastStreamId > -1) {
+            throw new IllegalArgumentException("Last stream ID must not be set on GOAWAY")
+          }
+          val lastStreamCreated = http2Handler.connection.remote.lastStreamCreated
+          val lastStreamId = (lastStreamCreated + goaway.extraStreamIds * 2) match {
+            case n if n < lastStreamCreated => Int.MaxValue
+            case n => n
+          }
+          http2Handler.goAway(
+            http2HandlerCtx,
+            lastStreamId,
+            goaway.errorCode,
+            goaway.content.retain(),
+            promise
+          ); ()
+        } finally { goaway.release(); () }
+
+      case data: Http2DataFrame =>
+        try {
+          http2Handler.encoder.writeData(
+            http2HandlerCtx,
+            data.streamId,
+            data.content.retain(),
+            data.padding,
+            data.isEndStream,
+            promise
+          ); ()
+        } finally { data.release(); () }
+
+      case headers: Http2HeadersFrame =>
+        http2Handler.encoder.writeHeaders(
+          http2HandlerCtx,
+          headers.streamId,
+          headers.headers(),
+          headers.padding,
+          headers.isEndStream,
+          promise
+        ); ()
+
+      case reset: Http2ResetFrame =>
+        http2Handler.resetStream(
+          http2HandlerCtx,
+          reset.streamId,
+          reset.errorCode,
+          promise
+        ); ()
+
+      case update: Http2WindowUpdateFrame =>
+        try {
+          http2Handler.connection.local.flowController.consumeBytes(
+            http2Handler.connection.stream(update.streamId),
+            update.windowSizeIncrement
+          )
+          promise.setSuccess(); ()
+        } catch {
+          case e: Throwable => promise.setFailure(e); ()
+        }
+
+      case msg => ctx.write(msg, promise); ()
+    }
+
+}
+
+object H2FrameCodec {
+
+  /** Aggressively acknowledge window updates */
+  val DefaultWindowUpdateRatio = 1.0f
```

Comment:
To make sure I understand: this means that any frames that are released will immediately trigger a window update?  Whereas a ratio of 0.95f would mean that releasing frames will not trigger a window update until 5% of the window size has been released?